c10/util/Exception.h

```diff
@@ -154,6 +154,12 @@ C10_API void set_warning_handler(WarningHandler* handler) noexcept(true);
 /// Gets the global warning handler.
 C10_API WarningHandler* get_warning_handler() noexcept(true);
 
+/// The TORCH_WARN_ONCE macro is difficult to test for. Use
+/// setWarnOnlyOnce(false) to turn it into TORCH_WARN, which can be
+/// tested for more easily.
+C10_API void set_warn_once(bool warn_once) noexcept(true);
```

Comment:
Overall the structure of this PR and the test look good. I have some questions about the context manager for testing and suggestions for the documentation, but I'm actually most concerned about the naming.

In particular, "set_warn_once" is ambiguous in its scope. That is: does it set all warnings to warn once? Maybe the flag can be named something like "always_warn", and then these functions would be named something like "set_always_warn" and "get_always_warn_enabled"? @kurtamohler, would that be consistent with our deterministic flag naming thinking?

Then this function would have the conceptual property: when always warn is False (default), some warnings will appear once and then be suppressed for the rest of the process; when always warn is True these warnings will not be suppressed. 

What are your thoughts @mattip?

Reply:
I am not really good with names. Your description makes sense to me.

Reply:
> I am not really good with names.

That makes two of us.

Reply:
I think I like the suggestion of changing `warn_once` to `always_warn`. I agree with the idea that it's a better conceptual model to have the default be False.

As for the consistency with deterministic flag names, `is_always_warn_enabled` is probably more consistent on the Python side. But there is also an argument for pairing a function that starts with `set_` with a function that starts with `get_`.

For the C++ side, I think consistency with other functions in `Exception.h` is more important than consistency with `Context::setDeterministicAlgorithms` and `Context::deterministicAlgorithms` in `Context.h`. As far as I can tell, the C++ side of this PR has the right consistency.